kubernetes_env/benchmark/benchmark.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+import argparse
+import logging
+import subprocess
+import os
+import sys
+import signal
+import requests
+import time
+import matplotlib.pyplot as plt
+import json
+from pathlib import Path
+
+sys.path.append("..")
+
+import kube_env
+import kube_util as util
+
+log = logging.getLogger(__name__)
+FILE_DIR = os.path.abspath(os.path.dirname(__file__))
+# Haven't found a workaround
+FORTIO_DIR = Path(__file__).parent.resolve().parent.resolve().parent.resolve()
+
+def plot(xaxis_data, yaxis_data, xaxis_label, yaxis_label, query, graph_name):
+    plt.plot(xaxis_data, yaxis_data, marker='o')
+    plt.xlabel(xaxis_label)
+    plt.ylabel(yaxis_label)
+    plt.title(query)
+    if not check_dir("graphs"):
+      os.mkdir(os.path.join(FILE_DIR, "graphs"))
+    plt.savefig(os.path.join(FILE_DIR, f"graphs/{graph_name}.png"))
+
+def check_dir(dirname):
+    return os.path.isdir(os.path.join(FILE_DIR, dirname))
+
+def make_dir(dirname):
+    os.mkdir(os.path.join(FILE_DIR, dirname))
+
+def run_fortio(platform, threads, qps, run_time):
+    if not check_dir("data"):
+      make_dir("data")
+    _, _, gateway_url = kube_env.get_gateway_info(platform)
+    cmd = f"{FORTIO_DIR}/bin/fortio "
+    cmd += f"load -c {threads} -qps {qps} -jitter -t {run_time}s -json data/output.json "
+    cmd += f"http://{gateway_url}/productpage"
+    fortio_proc = util.start_process(cmd, preexec_fn=os.setsid)
+    outs, errs = fortio_proc.communicate()
+
+def get_histogram(file_name):
+    with open(f"data/{file_name}", "r") as f:
+       fortio_json = json.load(f)
+       data = fortio_json["DurationHistogram"]
+       return data
+
+def benchmark_filter(platform, threads, qps, time, file_name, graph_name):
+    run_fortio(platform, threads, qps, time)
+    data = get_histogram(file_name)
+    if data is not None:
+      latency_data = data["Data"]
+      response_times = [resp["End"] - resp["Start"] for resp in latency_data]
+      request_counts = [resp["Count"] for resp in latency_data]
+      plot(response_times, request_counts, "Response time (ms)", 
+           "Requests", f"Platform: {platform} Threads: {threads} QPS: {qps}", graph_name)
+    else:
+      log.error("Error in processing data")
+
+def start_benchmark(filter_dirs, platform, 
+                    threads, qps, time, file_name, graph_name):
+    if kube_env.check_kubernetes_status() != util.EXIT_SUCCESS:
+      log.error("Kubernetes is not set up."
+                  " Did you run the deployment script?")
+      sys.exit(util.EXIT_FAILURE)
+
+    # Undeploy current running filter
+    kube_env.undeploy_filter() 
+    for (idx, filter_dir) in enumerate(filter_dirs):
```

Comment:
Are you going to supply the filter_dir? Or how is this intended to be used?

Reply:
Are you referring to line 75 or line 76?

For 75, I was thinking about undeploy whatever current running filter is for a fresh start but I think it's not necessary since it's kinda silly.

For 76, I was thinking users can supply list filter directory for the script to build, benchmark then undeploy

Reply:
Okay that makes sense. We are the user so we should do this with the next pull request.